src/htm/algorithms/SpatialPooler.cpp

```diff
@@ -860,34 +862,54 @@ void SpatialPooler::inhibitColumnsGlobal_(const vector<Real> &overlaps,
     overlaps_[i] += tieBreaker_[i];
 
   activeColumns.clear();
-  const UInt numDesired = (UInt)(density * numColumns_);
+  const UInt numDesired = static_cast<UInt>(density*numColumns_);
   NTA_CHECK(numDesired > 0) << "Not enough columns (" << numColumns_ << ") "
                             << "for desired density (" << density << ").";
   // Sort the columns by the amount of overlap.  First make a list of all of the
   // column indexes.
-  activeColumns.reserve(numColumns_);
-  for(UInt i = 0; i < numColumns_; i++)
-    activeColumns.push_back(i);
+  activeColumns.resize(numColumns_);
+  std::iota(activeColumns.begin(), activeColumns.end(), 0); //0,1,2,..,numColumns_-1
+
+  int same_overlap = 0;
   // Compare the column indexes by their overlap.
-  auto compare = [&overlaps_](const UInt &a, const UInt &b) -> bool
-    {return (overlaps_[a] == overlaps_[b]) ? a > b : overlaps_[a] > overlaps_[b];};
+  auto compare = [&overlaps_, &same_overlap](const UInt &a, const UInt &b) -> bool
+    {
+      if (overlaps_[a] == overlaps_[b]) {
+	same_overlap++;
+        return  a > b; //but we also need this for deterministic results
```

Comment:
problem 1: we are removing columns with the same overlap from the top n selection.